src/uploads/imageResizer.ts

```diff
@@ -6,128 +6,230 @@ import sharp from 'sharp';
+};
+
+/**
+ * Create a new image name based on the output image name, the dimensions and
+ * the extension.
+ *
+ * Ignore the fact that duplicate names could happen if the there is one
+ * size with `width AND height` and one with only `height OR width`. Because
+ * space is expensive, we will reuse the same image for both sizes.
+ *
+ * @param outputImageName - the sanitized image name
+ * @param bufferInfo - the buffer info
+ * @param extension - the extension to use
+ * @returns the new image name that is not taken
+ */
+const createImageName = (
+  outputImageName: string,
+  { width, height }: sharp.OutputInfo,
+  extension: string,
+) => `${outputImageName}-${width}x${height}.${extension}`;
+
+/**
+ * Create the result object for the image resize operation based on the
+ * provided parameters. If the name is not provided, an empty result object
+ * is returned.
+ *
+ * @param name - the name of the image
+ * @param filename - the filename of the image
+ * @param width - the width of the image
+ * @param height - the height of the image
+ * @param filesize - the filesize of the image
+ * @param mimeType - the mime type of the image
+ * @param sizesToSave - the sizes to save
+ * @returns the result object
+ */
+const createResult = (
+  name: string,
+  filename: FileSize['filename'] = null,
+  width: FileSize['width'] = null,
+  height: FileSize['height'] = null,
+  filesize: FileSize['filesize'] = null,
+  mimeType: FileSize['mimeType'] = null,
+  sizesToSave: FileToSave[] = [],
+): ImageSizesResult => ({
+  sizesToSave,
+  sizeData: {
+    [name]: {
+      width,
+      height,
+      filename,
+      filesize,
+      mimeType,
+    },
+  },
+});
+
+/**
+ * Check if the image needs to be resized according to the requested dimensions
+ * and the original image size. If the resize options withoutEnlargement or withoutReduction are provided,
+ * the image will be resized regardless of the requested dimensions, given that the
+ * width or height to be resized is provided.
+ *
+ * @param resizeConfig - object containing the requested dimensions and resize options
+ * @param original - the original image size
+ * @returns true if the image needs to be resized, false otherwise
+ */
+const needsResize = (
+  { width: desiredWidth, height: desiredHeigth, withoutEnlargement, withoutReduction }: ImageSize,
+  original: ProbedImageSize,
+): boolean => {
+  // allow enlargement or prevent reduction (our default is to prevent
+  // enlargement and allow reduction)
+  if (withoutEnlargement !== undefined || withoutReduction !== undefined) {
+    return true; // needs resize
+  }
+
+  const isWidthOrHeightNotDefined = !desiredHeigth || !desiredWidth;
+  if (isWidthOrHeightNotDefined) {
+    // If with and height are not defined, it means there is a format conversion
+    // and the image needs to be "resized" (transformed).
+    return true; // needs resize
+  }
+
+  const hasInsufficientWidth = original.width <= desiredWidth;
```

Comment:
@AlessioGr wouldn't that be the exact opposite? The original is already in the span of desired width and does not require any resizing.

`const hasSufficientWidth` would be better fitting in this case. 
The same applies for `hasSufficientHeight`.

Furthermore, with that logic, you will not resize any image that needs to be resized in height when the width is already sufficient.

Example: The image size desired is 640x640. The uploaded image is 600x1800. In that case, I would assume the image gets resized to 600x640 because of cropping.

Now the image will stay the same.

Reply:
Isn't the logic the same - just inverted?

Previously, it would return true (should resize) if the desired width <= original width.

Now, it returns false (shouldn't resize) if the original width <= desired width.

Just makes it a bit more clear as original width <= desired width would be what I expected to see when I think of "enlargement" or "sufficient width"

> Furthermore, with that logic, you will not resize any image that needs to be resized in height when the width is already sufficient.

You're completely right there, I missed that! Would it work if I change it to

if(hasInsufficientWidth && hasInsufficientHeight) { return false; }

while changing the <= to < for those two conditions?

Reply:
> Isn't the logic the same - just inverted?

Yes, that's why I propose to change the name to sufficient instead of insufficient.

> if(hasInsufficientWidth && hasInsufficientHeight) { return false; }
> while changing the <= to < for those two conditions?

Yes, this should do the trick.